File: tools/mcp-server/limn-mcp-server.pl
%% Limn MCP Server - Pure Prolog
%% =========================================================
%% fab mcp | pur prl | nu nod
%% *Building MCP. Pure Prolog. No Node.*
%%
%% The monk accepts no impurity.
%%
%% Usage: scryer-prolog limn-mcp-server.pl
%%        (reads JSON-RPC from stdin, writes to stdout)

:- use_module(library(lists)).
:- use_module(library(format)).
:- use_module(library(dcgs)).
:- use_module(library(charsio)).

%% Modules loaded via wrapper script

%% ============================================================
%% TOOL DEFINITIONS
%% ============================================================

mcp_tools([
    json([
        name-limn_interpret,
        description-'Interpret a Limn sentence. Words define REGIONS; sentences combine by INTERSECTION.',
        inputSchema-json([
            type-object,
            properties-json([
                sentence-json([type-string, description-'The Limn sentence']),
                key-json([type-string, description-'Optional context key'])
            ]),
            required-[sentence]
        ])
    ]),
    json([
        name-limn_compose,
        description-'Create Limn sentences from English concepts.',
        inputSchema-json([
            type-object,
            properties-json([
                concept-json([type-string, description-'English concept to express'])
            ]),
            required-[concept]
        ])
    ]),
    json([
        name-limn_validate,
        description-'Check if Limn sentence contains valid vocabulary.',
        inputSchema-json([
            type-object,
            properties-json([
                sentence-json([type-string, description-'Sentence to validate'])
            ]),
            required-[sentence]
        ])
    ]),
    json([
        name-limn_teach,
        description-'Get an interactive Limn teaching lesson.',
        inputSchema-json([
            type-object,
            properties-json([
                level-json([type-integer, description-'Lesson 1-3', default-1])
            ])
        ])
    ]),
    json([
        name-limn_poetry,
        description-'Generate Limn poetry on a theme.',
        inputSchema-json([
            type-object,
            properties-json([
                theme-json([type-string, description-'Theme for poetry'])
            ]),
            required-[theme]
        ])
    ]),
    json([
        name-limn_vocabulary,
        description-'Look up Limn words or browse by domain.',
        inputSchema-json([
            type-object,
            properties-json([
                word-json([type-string, description-'Word to look up']),
                domain-json([type-string, description-'Domain to browse'])
            ])
        ])
    ])
]).

%% ============================================================
%% TOOL HANDLERS
%% ============================================================

handle_tool(limn_validate, Args, Result) :-
    json_get(Args, sentence, Sentence),
    validate_sentence(Sentence, Valid, Invalid),
    length(Valid, VC),
    ( Invalid = [] ->
        number_chars(VC, VCChars),
        atom_chars(VCAtom, VCChars),
        atom_concat('Valid Limn. All ', VCAtom, T1),
        atom_concat(T1, ' words recognized.', Text)
    ;
        format(atom(Text), 'Unknown words: ~w', [Invalid])
    ),
    Result = json([content-[json([type-text, text-Text])]]).

handle_tool(limn_interpret, Args, Result) :-
    json_get(Args, sentence, Sentence),
    ( json_get(Args, key, Key) -> true ; Key = '' ),
    format(atom(Text), 'Interpret Limn: "~w"~nLimn words define REGIONS. Sentences combine by INTERSECTION.~nKey: ~w~nGenerate 5 interpretations.', [Sentence, Key]),
    Result = json([content-[json([type-text, text-Text])]]).

handle_tool(limn_compose, Args, Result) :-
    json_get(Args, concept, Concept),
    format(atom(Text), 'Compose Limn for: "~w"~n~nUse 3-letter words. Domains: physical (sol,liq,hot,col), mind (thi,fee,kno,dre), change (tra,mov,gro).~nOperators: nu (not), ve (very), | (scope).~nProvide 2-3 options.', [Concept]),
    Result = json([content-[json([type-text, text-Text])]]).

handle_tool(limn_teach, Args, Result) :-
    ( json_get(Args, level, Level) -> true ; Level = 1 ),
    lesson_text(Level, Text),
    Result = json([content-[json([type-text, text-Text])]]).

handle_tool(limn_poetry, Args, Result) :-
    json_get(Args, theme, Theme),
    format(atom(Text), 'Write Limn poetry on: "~w"~n~nUse 4-6 lines of valid Limn. Annotate each line.', [Theme]),
    Result = json([content-[json([type-text, text-Text])]]).

handle_tool(limn_vocabulary, Args, Result) :-
    ( json_get(Args, word, Word) ->
        ( word(Word) ->
            atom_concat(Word, ' is valid Limn vocabulary.', Text)
        ;
            atom_concat(Word, ' not found.', Text)
        )
    ; json_get(Args, domain, _) ->
        Text = 'Domains: physical, spatial, temporal, change, life, mind, communication, social, values.'
    ;
        Text = 'Use word param to look up, or domain to browse.'
    ),
    Result = json([content-[json([type-text, text-Text])]]).

%% Lesson texts
lesson_text(1, 'Lesson 1: Intersection. Words define REGIONS. hot col = lukewarm, contrast, ambivalence.').
lesson_text(2, 'Lesson 2: Order Independence. sol liq = liq sol = ice, slush, phase boundary.').
lesson_text(3, 'Lesson 3: Operators. nu negates NEXT word. nu sol liq vs sol nu liq.').
lesson_text(_, 'Lessons 1-3 available.').

%% ============================================================
%% JSON-RPC HANDLER
%% ============================================================

handle_request(Request, Response) :-
    json_get(Request, method, Method),
    ( json_get(Request, id, Id) -> true ; Id = null ),
    handle_method(Method, Request, Id, Response).

handle_method(initialize, _, Id, Response) :-
    Response = json([
        jsonrpc-'2.0',
        id-Id,
        result-json([
            protocolVersion-'2024-11-05',
            capabilities-json([tools-json([])]),
            serverInfo-json([name-'limn-mcp', version-'1.0.0'])
        ])
    ]).

handle_method('tools/list', _, Id, Response) :-
    mcp_tools(Tools),
    Response = json([
        jsonrpc-'2.0',
        id-Id,
        result-json([tools-Tools])
    ]).

handle_method('tools/call', Request, Id, Response) :-
    json_get(Request, params, Params),
    json_get(Params, name, ToolName),
    ( json_get(Params, arguments, Args) -> true ; Args = json([]) ),
    ( handle_tool(ToolName, Args, Result) ->
        Response = json([jsonrpc-'2.0', id-Id, result-Result])
    ;
        Response = json([jsonrpc-'2.0', id-Id, error-json([code-(-32603), message-'Tool error'])])
    ).

handle_method('notifications/initialized', _, _, _) :- !.  % No response

handle_method(Method, _, Id, Response) :-
    format(atom(Msg), 'Method not found: ~w', [Method]),
    Response = json([jsonrpc-'2.0', id-Id, error-json([code-(-32601), message-Msg])]).

%% ============================================================
%% MAIN LOOP
%% ============================================================

read_line(Chars) :-
    get_char(C),
    ( C = end_of_file -> Chars = []
    ; C = '\n' -> Chars = []
    ; Chars = [C|Rest], read_line(Rest)
    ).

main_loop :-
    read_line(Line),
    ( Line = [] -> true  % EOF
    ;
        ( json_parse(Line, Request) ->
            ( handle_request(Request, Response) ->
                json_write(Response, OutChars),
                atom_chars(Out, OutChars),
                write(Out), nl, flush_output
            ;
                true  % notification, no response
            )
        ;
            json_write(json([jsonrpc-'2.0', id-null, error-json([code-(-32700), message-'Parse error'])]), ErrChars),
            atom_chars(Err, ErrChars),
            write(Err), nl, flush_output
        ),
        main_loop
    ).

:- initialization(main_loop).

File: tools/mcp-server/limn-mcp-server.js
#!/usr/bin/env node
/**
 * Limn MCP Server - Node.js JSON-RPC wrapper
 * =========================================================
 * fab mcp | nod jso | prl lim
 * The bridge speaks JSON. The core speaks Limn.
 */

const { spawn } = require('child_process');
const path = require('path');
const readline = require('readline');

const LINTER_DIR = path.join(__dirname, '..', 'linter');

// Tool definitions
const TOOLS = [
  {
    name: "limn_interpret",
    description: "Interpret a Limn sentence. Words define REGIONS of meaning; sentences combine by INTERSECTION. Returns diverse interpretations, or collapsed meanings if a key is provided.",
    inputSchema: {
      type: "object",
      properties: {
        sentence: { type: "string", description: "The Limn sentence (e.g., 'sol liq tra')" },
        key: { type: "string", description: "Optional context key to collapse ambiguity (e.g., 'chemistry', 'emotions')" }
      },
      required: ["sentence"]
    }
  },
  {
    name: "limn_compose",
    description: "Create Limn sentences from English concepts. Returns multiple valid translations with explanations.",
    inputSchema: {
      type: "object",
      properties: {
        concept: { type: "string", description: "English concept to express (e.g., 'hidden truth slowly revealed')" }
      },
      required: ["concept"]
    }
  },
  {
    name: "limn_validate",
    description: "Check if a Limn sentence contains valid vocabulary words.",
    inputSchema: {
      type: "object",
      properties: {
        sentence: { type: "string", description: "The Limn sentence to validate" }
      },
      required: ["sentence"]
    }
  },
  {
    name: "limn_teach",
    description: "Get an interactive Limn teaching lesson.",
    inputSchema: {
      type: "object",
      properties: {
        level: { type: "integer", description: "Lesson level 1-3", default: 1 }
      }
    }
  },
  {
    name: "limn_poetry",
    description: "Generate Limn poetry on a given theme.",
    inputSchema: {
      type: "object",
      properties: {
        theme: { type: "string", description: "Theme for poetry (e.g., 'grief', 'transformation')" }
      },
      required: ["theme"]
    }
  },
  {
    name: "limn_vocabulary",
    description: "Look up Limn words or browse vocabulary by domain.",
    inputSchema: {
      type: "object",
      properties: {
        word: { type: "string", description: "Specific word to look up" },
        domain: { type: "string", description: "Domain to browse (physical, mind, change, etc.)" }
      }
    }
  }
];

// Run Prolog and get result
function runProlog(goal) {
  return new Promise((resolve, reject) => {
    const proc = spawn('scryer-prolog', [
      '-g', `consult('${LINTER_DIR}/limn-vocab.pl'), consult('limn-mcp.pl'), ${goal}, halt`
    ], { cwd: __dirname });

    let stdout = '';
    let stderr = '';

    proc.stdout.on('data', d => stdout += d);
    proc.stderr.on('data', d => stderr += d);
    proc.on('close', code => {
      if (code === 0) resolve(stdout);
      else reject(new Error(stderr || `Exit code ${code}`));
    });
    proc.on('error', reject);
  });
}

// Validate using Prolog linter
async function validateSentence(sentence) {
  try {
    const result = await runProlog(`handle_validate('${sentence.replace(/'/g, "''")}')`);
    const validMatch = result.match(/VALID_COUNT:(\d+)/);
    const invalidMatch = result.match(/INVALID_WORDS:(\[.*?\])/);

    const validCount = validMatch ? parseInt(validMatch[1]) : 0;
    const invalidWords = invalidMatch ? invalidMatch[1] : '[]';

    if (invalidWords === '[]') {
      return `✓ Valid Limn sentence. All ${validCount} words recognized.`;
    } else {
      return `⚠ Unknown words: ${invalidWords}\n\nCheck vocabulary or use similar concepts.`;
    }
  } catch (e) {
    return `Error validating: ${e.message}`;
  }
}

// Handle tool calls
async function handleTool(name, args) {
  switch (name) {
    case 'limn_validate':
      return await validateSentence(args.sentence || '');

    case 'limn_interpret': {
      const sentence = args.sentence || '';
      const key = args.key || '';
      let prompt = `Interpret this Limn sentence: "${sentence}"

Limn is a constructed language where:
- Words define REGIONS of meaning, not single referents
- Sentences combine words by INTERSECTION (where all meanings overlap)
- Word ORDER does not matter (commutative)
- Ambiguity is intentional until collapsed by a KEY

`;
      if (key) {
        prompt += `KEY PROVIDED: "${key}"

Collapse the ambiguity using this key. Provide 2-3 specific interpretations that fit this domain.`;
      } else {
        prompt += `NO KEY PROVIDED - Generate 5 diverse interpretations across different domains.

For each interpretation:
- Name the implicit key/domain
- Give the specific meaning
- Be creative but valid`;
      }
      return prompt;
    }

    case 'limn_compose': {
      const concept = args.concept || '';
      return `Compose a Limn sentence expressing: "${concept}"

Use ONLY valid Limn vocabulary (3-letter words):

PHYSICAL: sol (solid), liq (liquid), gas, hot, col (cold), bri (bright), dim
SPATIAL: abo (above), bel (below), ins (inside), out (outside), nea (near), far
TEMPORAL: now, pas (past), fut (future), beg (begin), end, cyc (cycle)
CHANGE: gro (growth), dec (decay), tra (transform), mov (movement), flo (flow)
LIFE: lif (life), dea (death), hea (health), you (young), old
MIND: thi (think), fee (feel), kno (know), dre (dream), cur (curious)
OPERATORS: nu (not), ve (very), so (somewhat), | (scope boundary)

Rules:
- Use 3-6 words typically
- Use | to separate contrasting concepts
- Be evocative, not explicit
- Provide 2-3 options with explanations`;
    }

    case 'limn_teach': {
      const level = args.level || 1;
      const lessons = {
        1: `## Limn Lesson 1: Basic Intersection

In Limn, words don't have single meanings - they define REGIONS.
When you combine words, their meanings INTERSECT.

**Exercise:** What could \`hot col\` mean?

This seems contradictory - hot AND cold. But think about what exists in that intersection...

Possible answers:
- Lukewarm (the temperature between)
- Thermal contrast (a situation where both exist)
- Temperature shock (rapid transition)
- Ambivalence (metaphorically, mixed feelings)

**Your turn:** What could \`bri dim\` mean?`,

        2: `## Limn Lesson 2: Order Independence

In Limn, word order doesn't matter - we're describing a REGION in meaning-space.

\`sol liq\` = \`liq sol\` (solid + liquid = liquid + solid)

Both mean: ice, glacier, slush, phase boundary

**Exercise:** These are ALL equivalent:
- \`lif gro you\`
- \`you gro lif\`
- \`gro lif you\`

All describe: alive + growing + young = seedling, child, new venture`,

        3: `## Limn Lesson 3: Operators

Operators modify the word that follows them.

\`nu\` = not/negation (applies to NEXT word only)

\`nu sol liq\` = (not-solid) AND liquid = emphasized liquidity
\`sol nu liq\` = solid AND (not-liquid) = dry solid, powder

**Exercise:** What's the difference between:
- \`nu lif gro\` = not-alive growth = machine process, crystal
- \`lif nu gro\` = alive not-growth = mature, stable, stasis`
      };
      return lessons[level] || 'Lessons 1-3 available. Specify level.';
    }

    case 'limn_poetry': {
      const theme = args.theme || 'transformation';
      return `Generate Limn poetry on the theme: "${theme}"

Create 4-6 lines of Limn with English annotations.
Use ONLY valid Limn vocabulary.

Format:
\`\`\`
[line 1 in limn]
[line 2 in limn]
[line 3 in limn]
[line 4 in limn]
\`\`\`

Then annotate each line:
**Line 1:** \`[limn]\` - [word breakdown] - [meaning]
**Line 2:** ...

End with a brief reflection on the poem's journey.`;
    }

    case 'limn_vocabulary': {
      const word = args.word;
      const domain = args.domain;

      if (word) {
        try {
          const result = await runProlog(`handle_vocab_word('${word}')`);
          if (result.includes('FOUND:true')) {
            return `✓ \`${word}\` is valid Limn vocabulary.`;
          } else {
            return `✗ \`${word}\` not found. Check spelling or browse domains.`;
          }
        } catch {
          return `Error looking up word.`;
        }
      } else if (domain) {
        const domains = {
          physical: 'sol liq gas hot col bri dim mag min',
          spatial: 'abo bel ins out nea far bet cen per',
          temporal: 'now pas fut beg end dur cyc mom ete',
          change: 'gro dec tra mov sta flo ris fal',
          life: 'lif dea hea sic you old bir see roo',
          mind: 'thi fee kno bel dou rem ima dre cur',
          communication: 'say lis wor mea amb cle que ans',
          social: 'sel oth fri ene joi sep giv tak',
          values: 'goo bad tru fal val zer one man all'
        };
        return domains[domain]
          ? `## ${domain}\n\n${domains[domain]}`
          : `Available domains: ${Object.keys(domains).join(', ')}`;
      } else {
        return `## Limn Vocabulary

**Domains:** physical, spatial, temporal, change, life, mind, communication, social, values

**Operators:** nu (not), ve (very), so (somewhat), te (tentative), we (weak)
**Quantifiers:** al (all), ex (exists), on (one)
**References:** yo (this), an (that), sa (same)

Use \`word\` param to look up specific words, or \`domain\` to browse.`;
      }
    }

    default:
      throw new Error(`Unknown tool: ${name}`);
  }
}

// Main MCP server loop
const rl = readline.createInterface({ input: process.stdin, output: process.stdout, terminal: false });

rl.on('line', async (line) => {
  try {
    const request = JSON.parse(line);
    const { method, id, params } = request;

    let result;

    switch (method) {
      case 'initialize':
        result = {
          protocolVersion: "2024-11-05",
          capabilities: { tools: {} },
          serverInfo: { name: "limn-mcp", version: "1.0.0" }
        };
        break;

      case 'tools/list':
        result = { tools: TOOLS };
        break;

      case 'tools/call':
        const content = await handleTool(params.name, params.arguments || {});
        result = { content: [{ type: "text", text: content }] };
        break;

      case 'notifications/initialized':
        return; // No response for notifications

      default:
        if (id !== undefined) {
          console.log(JSON.stringify({
            jsonrpc: "2.0",
            id,
            error: { code: -32601, message: `Method not found: ${method}` }
          }));
        }
        return;
    }

    if (id !== undefined) {
      console.log(JSON.stringify({ jsonrpc: "2.0", id, result }));
    }

  } catch (e) {
    console.log(JSON.stringify({
      jsonrpc: "2.0",
      id: null,
      error: { code: -32603, message: e.message }
    }));
  }
});
